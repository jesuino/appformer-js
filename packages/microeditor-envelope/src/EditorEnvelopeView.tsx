import * as React from "react";
import * as ReactDOM from "react-dom";
import * as AppFormer from "appformer-js-core";
import { LoadingScreen } from "./LoadingScreen";

interface Props {
  exposing: (self: EditorEnvelopeView) => void;
  loadingScreenContainer: HTMLElement;
}

interface State {
  editor?: AppFormer.Editor;
  loading: boolean;
}

export class EditorEnvelopeView extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = { editor: undefined, loading: true };
    this.props.exposing(this);
  }

  public getEditor() {
    return this.state.editor;
  }

  public setEditor(editor: AppFormer.Editor) {
    return new Promise(res => this.setState({ editor: editor }, res));
  }

  public setLoadingFinished() {
    return new Promise(res => this.setState({ loading: false }, res));
  }

  private LoadingScreenPortal() {
    return ReactDOM.createPortal(<LoadingScreen visible={this.state.loading} />, this.props.loadingScreenContainer!);
  }

  public render() {
    return (
      <>
        {this.LoadingScreenPortal()}
        {this.state.editor && this.state.editor.af_isReact && this.state.editor.af_componentRoot()}
      </>
    );
  }
}
