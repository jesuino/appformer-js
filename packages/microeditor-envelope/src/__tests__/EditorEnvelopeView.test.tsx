import * as React from "react";
import { shallow } from "enzyme";
import { EditorEnvelopeView } from "../EditorEnvelopeView";
import { DummyEditor } from "./DummyEditor";

let loadingScreenContainer: HTMLElement;
beforeEach(() => (loadingScreenContainer = document.body.appendChild(document.createElement("div"))));
afterEach(() => loadingScreenContainer.remove());

function renderEditorEnvelopeView(): [EditorEnvelopeView, ReturnType<typeof shallow>] {
  let view: EditorEnvelopeView;
  const render = shallow(
    <EditorEnvelopeView exposing={self => (view = self)} loadingScreenContainer={loadingScreenContainer} />
  );
  return [view!, render];
}

describe("EditorEnvelopeView", () => {
  test("first open", () => {
    const [_, render] = renderEditorEnvelopeView();
    expect(render).toMatchSnapshot();
  });

  test("after loading stops", async () => {
    const [view, render] = renderEditorEnvelopeView();
    await view.setLoadingFinished();
    expect(render).toMatchSnapshot();
  });

  test("after loading stops and editor is set", async () => {
    const [view, render] = renderEditorEnvelopeView();
    await view.setLoadingFinished();
    await view.setEditor(new DummyEditor());
    expect(render).toMatchSnapshot();
  });
});
