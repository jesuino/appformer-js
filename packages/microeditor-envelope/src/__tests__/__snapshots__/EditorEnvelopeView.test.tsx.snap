// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorEnvelopeView after loading stops 1`] = `
<Fragment>
  <Portal
    containerInfo={<div />}
  >
    <LoadingScreen
      visible={false}
    />
  </Portal>
</Fragment>
`;

exports[`EditorEnvelopeView after loading stops and editor is set 1`] = `
<Fragment>
  <Portal
    containerInfo={<div />}
  >
    <LoadingScreen
      visible={false}
    />
  </Portal>
  <div>
    This is a dummy editor!
  </div>
</Fragment>
`;

exports[`EditorEnvelopeView first open 1`] = `
<Fragment>
  <Portal
    containerInfo={<div />}
  >
    <LoadingScreen
      visible={true}
    />
  </Portal>
</Fragment>
`;
